### ISSUE TITLE
[Bug]: Error occurred when attempting to extract image dimensions: #<FastImage::UnknownImageType: FastImage::UnknownImageType>

### ISSUE THREAD

**Issue by dylanfisher** [shrinerb/shrine#609]
### Report

I've recently updated an app's Gemfile, and now receive the error `Error occurred when attempting to extract image dimensions: #<FastImage::UnknownImageType: FastImage::UnknownImageType>` when uploading a jpeg image. Apologies in advance if this is not the right repo to file the issue, but I've narrowed it down to one line in the Shrine code that may or may not depend on another gem.

I've narrowed down the gem update that causes the issue in Shrine. Upgrading the `net-protocol` gem from `gem "net-protocol", "0.1.3"` to version `gem "net-protocol", "0.2.0"` causes the following problem at this line in the `store_dimensions` plugin https://github.com/shrinerb/shrine/blob/76f8aecac72fbb65573f9494e6fbda6f83a71a2f/lib/shrine/plugins/store_dimensions.rb#L120. 

If I debug that point, I find the following:

```ruby
plugin :store_dimensions, analyzer: -> (io, analyzers) do
  FastImage.size(io, raise_on_failure: true)
  # => FastImage::UnknownImageType: FastImage::UnknownImageType
end
```

But if I call Fastimage on the `io.url` object, it does work:

```ruby
plugin :store_dimensions, analyzer: -> (io, analyzers) do
  FastImage.size(io.url, raise_on_failure: true)
  # => [1500, 1000]
end
```

### Expected Behavior

`store_dimensions` Fastimage plugin analyzer should extract dimensions of `io` object.

### Actual Behavior

Unrelated gem update causes issue in `store_dimensions` Fastimage plugin analyzer.

### Steps to Reproduce the Problem

1. Use gem `"net-protocol", "0.2.0"`
2. Upload jpeg image with `store_dimensions` plugin and `Fastimage` analyzer.

### Ruby Version

3.1.3

### Shrine Version

3.4.0

### Anything else?

`gem "rails", "~> 7.0.1"`
`gem "net-protocol", "0.2.0"`

**Comment by dylanfisher** [comment]
The `net-protocol` 0.2.0 release tag is here https://github.com/ruby/net-protocol/releases/tag/v0.2.0

**Comment by dylanfisher** [comment]
S3 storage https://github.com/dylanfisher/forest/blob/3c90b1c0bdc64cadf59464f34735fba89b965863/lib/forest/shrine.rb

**Comment by janko** [comment]
This is the only commit in 0.2.0 release – https://github.com/ruby/net-protocol/commit/781e400389e75284327d97a3185d9c05537f2611 – and it does seem to change a lot of behaviour, so it's very possible it introduced a breaking change. aws-sdk-s3 uses net/http for HTTP requests, so it's possible streaming is not working correctly anymore. Someone would have to isolate the problem, ideally using net/http directly – no Shrine or aws-sdk-s3 or Down – and report it to `ruby/net-protocol`.

**Comment by janko** [comment]
OK, it seems that newer `net-protocol` reuses string objects when yielding chunks. Previously, in the code below, yielded chunks were all different ruby objects, and now they're all the same object:

```rb
client = Aws::S3::Client.new(...)
client.put_object(bucket: "my-bucket", key: "foo", body: "a" * 100_000)

object_ids = []
client.get_object(bucket: "my-bucket", key: "foo") do |chunk|
  object_ids << chunk.object_id
end

object_ids #=>
# net-protocol 0.1.3 – [5100, 5120, 5140, 5160, 5180, 5200, 5220, 5240, 5260, 5280, 5300]
# net-protocol 0.2.0 – [5100, 5100, 5100, 5100, 5100, 5100, 5100, 5100, 5100, 5100]
```

This doesn't cause issues in aws-sdk-s3 itself, but Shrine S3 storage uses `Down::ChunkedIO` from the Down gem to wrap the streaming, which relies on these chunks being different objects. 

It does seem like it was an intentional behavior change in net-protocol, to make streaming an HTTP response more memory-efficient, but I would consider it a breaking change. Technically, net-protocol is in the `0.x` phase, so AFAIK minor versions are still allowed to introduce breaking changes.

**Comment by jrochkind** [comment]
Nice find. Interesting -- so the same String object, but with mutated contents?

I can see how that would be a lot more performant. My apps deal with streaming very large (~200-500 megabyte) files, via shrine and down, and I have noticed that there can be pretty severe memory and performance (from GC) costs to doing so, which has been disappointing to me. 

I guess this change could be towards amelioerating that... depending on what Down::ChunkedIO goes on to do with it, maybe. 

I wouldn't necessarily have expected Down::ChunkedIO to rely on the chunk being differnet objects; presumably the net-protocol authors didn't expect anything to either, or they would have warned more clearly about the change. Or not done it -- although the performance improvements really will be welcome. (That a part of the stdlib is still at pre-1.0 and could release breaking changes with no notice is not awesome either :( )

Does Down actually fail it's CI tests when using net-protocol 0.2.0? That'd be a nice signal of reliability there, and make things easier! When trying to fix Down::ChunkedIO, whenever someone gets around to it -- it would be nice to see if there's a way to preserve and continue the performance enhancements here -- ruby really does pay a penalty when creating lots of objects (RAM growth that can't always be reversed; GC time), so net-protocol reducing that would be a boon -- but not if Down::ChunkedIO just (eg) calls `dup` on em all to make just as many new objects as before!

**Comment by janko** [comment]
This issue should be resolved by upgrading to net-protocol 0.2.1.
